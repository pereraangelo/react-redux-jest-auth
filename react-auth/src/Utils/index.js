import moment from "moment";

export const getGreetingTime = () => {
  let currentTime = new Date();

  const splitAfternoon = 12; // 24hr time to split the afternoon
  const splitEvening = 15; // 24hr time to split the evening
  const currentHour = moment(currentTime).format("HH");

  if (currentHour >= splitAfternoon && currentHour <= splitEvening) {
    // Between 12 PM and 5PM
    return "Good Afternoon";
  } else if (currentHour >= splitEvening) {
    // Between 5PM and Midnight
    return "Good Evening";
  }
  // Between Dawn and noon
  return "Good Morning";
};



export const findByTestAtrr = (component, attr) => {
  const wrapper = component.find(`[data-test='${attr}']`);
  return wrapper;
};

export const simulateChangeOnInput = (wrapper, inputSelector, newValue) => {
  let input = findByTestAtrr(wrapper, inputSelector);
  input.simulate("change", {
    target: { name: inputSelector, value: newValue },
  });

  return findByTestAtrr(wrapper, inputSelector);
};


export const simulateCheckBox = (wrapper, inputSelector, newValue) => {
    let input = findByTestAtrr(wrapper, inputSelector);
    input.simulate('change', {target: {name:inputSelector,checked: newValue}});
  
    return findByTestAtrr(wrapper, inputSelector);
  };

  