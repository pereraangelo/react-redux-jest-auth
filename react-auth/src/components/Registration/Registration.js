import React, {  useState } from "react";
import * as actions from "../../actions/authActions";
import { connect } from "react-redux";
import { Link } from "react-router-dom";
import { Alert } from "reactstrap";
 
export function Registration(props) {
  const { register, reset, alert, potentialUser, alreadyAuser } = props;

  const [user, setUser] = useState({
    userName: "",
    email: "",
    password: "",
  });

  const [showPass, setShowPass] = useState(false);
  const [acceptTerms, setAcceptTerms] = useState(false);
  const [isSubmitted, setIsSubmitted] = useState(false);
  const [confirmPassword, setConfirmPassword] = useState("");

  /*----------- input handler to set input values-----------*/
  const handleInput = (event) => {
    const target = event.target;
    const value = target.value;
    const name = target.name;
    setUser((preState) => ({ ...preState, [name]: value }));
  };

  /*----------- submit the loginform to login -----------*/
  const handleSubmit = () => {
    setIsSubmitted(true);
    if (validateEmail() && user.userName && acceptTerms && user.password===confirmPassword) {
      register(user);
      document.getElementById("registrationForm").reset();
      setShowPass(false);
      setAcceptTerms(false);
      setIsSubmitted(false);
      setConfirmPassword("");
      setUser({
        userName: "",
        email: "",
        password: "",
      });
    }
  };


/*----------- validity check for email input -----------*/
  const validateEmail = () => {
    var re =
      /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
    return re.test(user.email);
  };

  /*----------- accept terms and conditions -----------*/
  const terms = (e) => {
    if (e.target.checked) {
      setAcceptTerms(true);
    } else {
      setAcceptTerms(false);
    }
  };

    /*----------- show hid password -----------*/
  const togglePass = () => {
    setShowPass(!showPass);
  };

 

  return (
    <div className="center">
      <div className="register card">
        <h3 className="card-header  login-header">Register</h3>
        <div className="card-body">
          <form id="registrationForm">
            <div className="form-group">
              <input
                data-test="userName"
                type="text"
                value={user.userName}
                name="userName"
                placeholder="User Name"
                onChange={handleInput}
                className={
                  "form-control" +
                  (isSubmitted && !user.userName ? " is-invalid" : "")
                }
              />
              {isSubmitted && !user.userName ? (
                <span className="error ">
                  <i
                    className="fa fa-exclamation-triangle"
                    aria-hidden="true"
                  ></i>
                  This field is required.
                </span>
              ) : (
                <span></span>
              )}
            </div>

            <div className="form-group">
              <input
                data-test="email"
                type="text"
                name="email"
                value={potentialUser || user.email}
                // defaultValue={potentialUser || ""}
                placeholder="Email"
                onChange={handleInput}
                className={
                  "form-control" +
                  (isSubmitted && (!validateEmail() || !user.email)
                    ? " is-invalid"
                    : "")
                }
              />
              {isSubmitted && (!validateEmail() || !user.email) ? (
                <span className="error ">
                  <i
                    className="fa fa-exclamation-triangle"
                    aria-hidden="true"
                  ></i>
                  Email is not valid
                </span>
              ) : (
                <span></span>
              )}
            </div>

            <div className="form-row">
              <div className="form-group col">
                <input
                  data-test="password"
                  type={showPass ? "text" : "password"}
                  name="password"
                  value={user.password}
                  placeholder="Enter Password"
                  onChange={handleInput}
                  className={
                    "form-control" +
                    (isSubmitted && !user.password ? " is-invalid" : "")
                  }
                />
                {isSubmitted && !user.password ? (
                  <span className="error">
                    <i
                      className="fa fa-exclamation-triangle"
                      aria-hidden="true"
                    ></i>
                    This field is required.
                    <br />
                  </span>
                ) : (
                  <span></span>
                )}
              </div>
              <div className="form-group col showPass">
                <input
                  data-test="confirmPassword"
                  type={showPass ? "text" : "password"}
                  name="confirmPassword"
                  placeholder="Confirm Password"
                  value={confirmPassword}
                  onChange={(e) => setConfirmPassword(e.target.value)}
                  className={
                    "form-control" +
                    (user.password !== confirmPassword ? " is-invalid" : "")
                  }
                />

                {user.password !== confirmPassword ? (
                  <span className="error">
                    <i
                      className="fa fa-exclamation-triangle "
                      aria-hidden="true"
                    ></i>
                    Password missmatch
                    <br />
                  </span>
                ) : (
                  <span></span>
                )}
              </div>
            </div>
            <div className="form-group form-check mt-2">
              <input
                data-test="showPassword"
                id="showPassword"
                type="checkbox"
                value={showPass}
                name="showPassword"
                onChange={togglePass}
                className={"form-check-input "}
              />
              <label htmlFor="showPassword" className="form-check-label">
                Show password
              </label>
            </div>

            <div className="form-group form-check mt-2">
              <input
                data-test="acceptTerms"
                id="acceptTerms"
                type="checkbox"
                value={acceptTerms}
                name="acceptTerms"
                onChange={terms}
                className={
                  "form-check-input " +
                  (!acceptTerms && isSubmitted ? " is-invalid" : "")
                }
              />
              <label htmlFor="acceptTerms" className="form-check-label">
                Accept Terms & Conditions
              </label>
            </div>

            <div className="form-group">
              <div className="form-group col d-flex justify-content-center align-items-center">
                <button
                  className="myBtn btn btn-primary"
                  type="button"
                  value="Register"
                  onClick={handleSubmit}
                  data-test="register"
                >
                  {" "}
                  <span className="skew-fix">Register</span>{" "}
                </button>

                <Link to="/" className=" btn-link pr-0" data-test="cancle">
                  Cancle
                </Link>
              </div>
            </div>

            {alert && (
              <div className="form-row" data-test="alert-element">
                <div className="form-group col ">
                  <Alert>
                    {alert}
                    {alreadyAuser && (
                      <Link to="/" className=" btn-link pr-0" data-test="login">
                        Login
                      </Link>
                    )}
                  </Alert>
                </div>
              </div>
            )}
          </form>
        </div>
      </div>
      <Link
        to="/FAQ"
        className=" faq-button btn-link pr-0"
        data-test="FAQ"
      ></Link>
    </div>
  );
}
const mapStateToProps = (state) => {
  return {
    alert: state.auth.alert,
    potentialUser: state.auth.potentialUser,
    alreadyAuser: state.auth.alreadyAuser,
  };
};

export default connect(mapStateToProps, actions)(Registration);
