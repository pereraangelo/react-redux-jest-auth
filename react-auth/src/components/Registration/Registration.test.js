import React from 'react'
import ReactDOM from 'react-dom'
import { shallow} from 'enzyme'
import renderer from 'react-test-renderer'
import {MemoryRouter} from "react-router"
import  {Registration} from './Registration'
import { findByTestAtrr, simulateChangeOnInput, simulateCheckBox } from '../../Utils'
 
const defaultProps={
    register: jest.fn(),
    alert:"",
    potentialUser:"",
    alreadyAuser:""
}
 

describe('Register Component',()=>{
    const wrapper = shallow(<Registration {...defaultProps} />);

    beforeEach(()=>jest.resetAllMocks())

    it('Register page should load without crahing',()=>{
        const div = document.createElement('div');
        ReactDOM.render(<MemoryRouter ><Registration/></MemoryRouter>, div);
    });

    it('Register component matches the snapshot',()=>{
        const tree=renderer.create(<MemoryRouter ><Registration/></MemoryRouter>).toJSON()
        expect(tree).toMatchSnapshot()
    });

    it('check form contains 5 inputs, 2 button,2 links,1 checkbox ',()=>{
        expect(wrapper.find('form').length).toBe(1)
        expect(wrapper.find('input').length).toBe(6);
        expect(wrapper.find('button').length).toBe(1);
        expect(wrapper.find('Link').length).toBe(2);


        expect(findByTestAtrr(wrapper, 'userName').length).toEqual(1)
        expect(findByTestAtrr(wrapper, 'email').length).toEqual(1)
        expect(findByTestAtrr(wrapper, 'password').length).toEqual(1)
        expect(findByTestAtrr(wrapper, 'confirmPassword').length).toEqual(1)
        expect(findByTestAtrr(wrapper, 'acceptTerms').length).toEqual(1)
        expect(findByTestAtrr(wrapper, 'showPassword').length).toEqual(1)
        expect(findByTestAtrr(wrapper, 'register').length).toBe(1); 
        expect(findByTestAtrr(wrapper, 'cancle').length).toBe(1); 
        expect(findByTestAtrr(wrapper, 'FAQ').length).toBe(1); 
        expect(findByTestAtrr(wrapper, 'login')).toBeTruthy(); 
        
    })



    it('let fill the form',()=>{

        const userNameInput         =   simulateChangeOnInput(wrapper,'userName','John')
        const emailInput            =   simulateChangeOnInput(wrapper,'email','john@gmail.com')
        const passwordInput         =   simulateChangeOnInput(wrapper,'password','pass123')
        const confirmPasswordInput  =   simulateChangeOnInput(wrapper,'confirmPassword','pass123')
        const acceptAggrement       =   simulateCheckBox(wrapper,'acceptTerms',true)

        expect(userNameInput.props().value).toEqual('John')
        expect(emailInput.props().value).toEqual('john@gmail.com')
        expect(passwordInput.props().value).toEqual('pass123')
        expect(confirmPasswordInput.props().value).toEqual('pass123')
        expect(acceptAggrement.props().value).toEqual(true)
    
    })



    it('fills email field if user enter it on login page ',()=>{

        wrapper.setProps({ potentialUser: 'john@gmail.com' });
        const testdata=findByTestAtrr(wrapper, 'email').props().value
        expect(testdata).toEqual('john@gmail.com')

    })



    it('shows password in plain text',()=>{

        simulateCheckBox(wrapper,'showPassword',true)
        const passwordField = findByTestAtrr(wrapper, 'password').props().type
        const confirmPasswordField = findByTestAtrr(wrapper, 'password').props().type

        expect(passwordField).toEqual("text")
        expect(confirmPasswordField).toEqual("text")
    })



 
    it('redirect user to Login page',()=>{
        const cancle = findByTestAtrr(wrapper, 'cancle').props().to
        
        expect(cancle).toEqual("/")
         
    })



    it('redirect user to faq page',()=>{
        const faq = findByTestAtrr(wrapper, 'FAQ').props().to
        expect(faq).toEqual("/FAQ")
         
    })



    it('shows alert if user clicks the registration button ',()=>{

        wrapper.setProps({ alert: 'test alert' });
        const testdata=findByTestAtrr(wrapper, 'alert-element').length
        expect(testdata).toEqual(1)
 
    })


    
    it('shows alert with link back to login page, if user alerady registered ',()=>{

        wrapper.setProps({ alert: 'test alert' });
        wrapper.setProps({ alreadyAuser: true });
        const testdata=findByTestAtrr(wrapper, 'login').props().to
        expect(testdata).toEqual("/")
 
    })

})

 
 