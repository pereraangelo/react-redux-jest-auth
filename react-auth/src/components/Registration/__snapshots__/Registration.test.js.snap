// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register Component Register component matches the snapshot 1`] = `
<div
  className="center"
>
  <div
    className="register card"
  >
    <h3
      className="card-header  login-header"
    >
      Register
    </h3>
    <div
      className="card-body"
    >
      <form
        id="registrationForm"
      >
        <div
          className="form-group"
        >
          <input
            className="form-control"
            data-test="userName"
            name="userName"
            onChange={[Function]}
            placeholder="User Name"
            type="text"
            value=""
          />
          <span />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            data-test="email"
            name="email"
            onChange={[Function]}
            placeholder="Email"
            type="text"
            value=""
          />
          <span />
        </div>
        <div
          className="form-row"
        >
          <div
            className="form-group col"
          >
            <input
              className="form-control"
              data-test="password"
              name="password"
              onChange={[Function]}
              placeholder="Enter Password"
              type="password"
              value=""
            />
            <span />
          </div>
          <div
            className="form-group col showPass"
          >
            <input
              className="form-control"
              data-test="confirmPassword"
              name="confirmPassword"
              onChange={[Function]}
              placeholder="Confirm Password"
              type="password"
              value=""
            />
            <span />
          </div>
        </div>
        <div
          className="form-group form-check mt-2"
        >
          <input
            className="form-check-input "
            data-test="showPassword"
            id="showPassword"
            name="showPassword"
            onChange={[Function]}
            type="checkbox"
            value={false}
          />
          <label
            className="form-check-label"
            htmlFor="showPassword"
          >
            Show password
          </label>
        </div>
        <div
          className="form-group form-check mt-2"
        >
          <input
            className="form-check-input "
            data-test="acceptTerms"
            id="acceptTerms"
            name="acceptTerms"
            onChange={[Function]}
            type="checkbox"
            value={false}
          />
          <label
            className="form-check-label"
            htmlFor="acceptTerms"
          >
            Accept Terms & Conditions
          </label>
        </div>
        <div
          className="form-group"
        >
          <div
            className="form-group col d-flex justify-content-center align-items-center"
          >
            <button
              className="myBtn btn btn-primary"
              data-test="register"
              onClick={[Function]}
              type="button"
              value="Register"
            >
               
              <span
                className="skew-fix"
              >
                Register
              </span>
               
            </button>
            <a
              className=" btn-link pr-0"
              data-test="cancle"
              href="/"
              onClick={[Function]}
            >
              Cancle
            </a>
          </div>
        </div>
      </form>
    </div>
  </div>
  <a
    className=" faq-button btn-link pr-0"
    data-test="FAQ"
    href="/FAQ"
    onClick={[Function]}
  />
</div>
`;
